File: ios/Classes/GfmcFlutterPlugin.swift
import Flutter
import JessicaSDK
import UIKit

/// Dart-facing entry point. Registered automatically via pubspec.yaml's
/// `flutter.plugin.platforms.ios.pluginClass` — hosts don't touch this class
/// directly, only lib/gfmc_flutter.dart's GfmcSdk facade.
///
/// Implements GfmcHostApi (Dart -> native calls) directly, and owns a
/// GfmcFlutterApi instance (native -> Dart calls) built once attached to an
/// engine. Wraps JessicaSDK (https://github.com/BDN-ID/gfmc-ios) — see that
/// package's README for the native API this mirrors; mirrors
/// GfmcFlutterPlugin.kt on the Android side field-for-field where the two
/// SDKs' surfaces line up.
public class GfmcFlutterPlugin: NSObject, FlutterPlugin, GfmcHostApi {

  private var flutterApi: GfmcFlutterApi?

  // Non-nil once GfmcHostApi.init() has actually been called via [init]
  // below. Guards against skuListener/tokenRefresher/delegate callbacks
  // firing into a torn-down engine after detach.
  private var attached = false

  // Jessica.shared has no public getter for the config it was last
  // configure()'d with (same situation as GfmcSDK.getConfig() being
  // `internal` on Android — see GfmcFlutterPlugin.kt's comment). Track the
  // last config passed to [init] ourselves instead.
  private var lastConfig: GfmcConfigMessage?

  // The hub JessicaSDK.open() handed back, so [closeMiniApp] has something
  // to dismiss. Android has no such hook (open() there returns Unit) —
  // that's the "no-op placeholder" gap noted in Messages.g.kt; iOS doesn't
  // share it, since JessicaHubViewController is a real, dismissable
  // UIViewController.
  private weak var hubViewController: JessicaHubViewController?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = GfmcFlutterPlugin()
    instance.flutterApi = GfmcFlutterApi(binaryMessenger: registrar.messenger())
    GfmcHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    instance.attached = true
  }

  // -- GfmcHostApi (Dart -> native) ------------------------------------------

  // JessicaSDK's `Jessica` class is @MainActor-isolated end to end (see its
  // .swiftinterface). GfmcHostApi's methods are plain nonisolated protocol
  // requirements -- Pigeon has no notion of actor isolation -- but Flutter
  // guarantees platform-channel handlers run on the platform thread, which
  // on iOS *is* the main thread, so MainActor.assumeIsolated is a correct
  // (not a hopeful) escape hatch here, not a data-race risk.

  public func `init`(config: GfmcConfigMessage) throws {
    MainActor.assumeIsolated {
      let sdkConfig = JessicaSDKConfig(
        environment: config.environment.toSdkEnv(),
        locale: config.locale,
        theme: config.theme.toSdkTheme(),
        isLoggingEnabled: config.enableLogging,
        connectionTimeout: TimeInterval(config.connectionTimeoutMs) / 1000.0
      )

      Jessica.shared.configure(sdkConfig)
      Jessica.shared.setDelegate(self)
      Jessica.shared.setSKUListener(skuListener)
      Jessica.shared.setTokenRefresher(tokenRefresher)
    }
    lastConfig = config
  }

  public func open(jwt: String) throws {
    MainActor.assumeIsolated {
      let presenter = topViewController()
      hubViewController = Jessica.shared.open(from: presenter, jwt: jwt)
    }
  }

  public func closeMiniApp() throws {
    // Mirrors GfmcSDK.closeMiniApp() on Android in effect (see this file's
    // hubViewController comment for why the two sides differ underneath).
    MainActor.assumeIsolated {
      hubViewController?.dismiss(animated: true)
    }
    hubViewController = nil
  }

  public func getVersion() throws -> GfmcVersionMessage {
    let v = MainActor.assumeIsolated { Jessica.version }
    return GfmcVersionMessage(
      name: v.name,
      versionCode: Int64(v.build),
      artifactVersion: JessicaSDKVersion.packageVersion,
      displayName: v.displayName
    )
  }

  public func getConfig() throws -> GfmcConfigMessage? {
    return lastConfig
  }

  // -- Presenting view controller -------------------------------------------

  // JessicaHubViewController is presented modally, so JessicaSDK needs
  // *something* to present it from. There's no ActivityAware equivalent on
  // iOS to hand this a live UIViewController the way GfmcFlutterPlugin.kt
  // gets one from ActivityPluginBinding — walk the key window's presented
  // controller chain instead, same approach most Flutter plugins that
  // present UIKit view controllers use (e.g. image_picker, url_launcher_ios).
  @MainActor
  private func topViewController() -> UIViewController? {
    let keyWindow = UIApplication.shared.connectedScenes
      .compactMap { ($0 as? UIWindowScene)?.keyWindow }
      .first
    var top = keyWindow?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }

  // -- JessicaSDK callbacks -> GfmcFlutterApi (native -> Dart) ---------------

  private lazy var skuListener: JessicaSKUListener = { [weak self] sku in
    guard let self, self.attached else { return }
    self.flutterApi?.onSkuSelected(sku: sku) { _ in }
  }

  private lazy var tokenRefresher: JessicaTokenRefresher = { [weak self] done in
    guard let self, self.attached, let api = self.flutterApi else {
      done(.failure(JessicaRefreshError("Flutter engine detached")))
      return
    }
    api.refreshToken { result in
      switch result {
      case .success(let jwt):
        done(.success(jwt))
      case .failure(let error):
        done(.failure(JessicaRefreshError(error.message ?? "refresh failed")))
      }
    }
  }
}

// -- JessicaSDKDelegate -----------------------------------------------------

extension GfmcFlutterPlugin: JessicaSDKDelegate {
  public func jessicaHubDidBecomeReady() {
    guard attached else { return }
    flutterApi?.onHubReady { _ in }
  }

  public func jessicaHubDidClose() {
    guard attached else { return }
    hubViewController = nil
    flutterApi?.onHubClosed { _ in }
  }

  public func jessicaHub(didRequestShare url: URL, title: String) {
    guard attached else { return }
    flutterApi?.onShareRequested(url: url.absoluteString, title: title) { _ in }
  }

  public func jessicaHub(didFailWith error: JessicaSDKError, message: String) {
    guard attached else { return }
    flutterApi?.onError(code: error.toMessage(), message: message) { _ in }
  }

  public func jessicaHub(didChangeModule module: JessicaModule) {
    guard attached else { return }
    flutterApi?.onModuleChanged(module: module.toMessage()) { _ in }
  }

  public func jessicaHub(didCompletePurchase sku: String, transactionId: String) {
    guard attached else { return }
    flutterApi?.onPurchaseCompleted(sku: sku, txId: transactionId) { _ in }
  }

  public func jessicaHub(didFailPurchase reason: String) {
    guard attached else { return }
    flutterApi?.onPurchaseFailed(reason: reason) { _ in }
  }
}

// -- Enum conversions ---------------------------------------------------

private extension GfmcEnvMessage {
  func toSdkEnv() -> JessicaEnvironment {
    switch self {
    case .production: return .production
    case .sandbox: return .sandbox
    case .dev: return .dev
    }
  }
}

private extension GfmcThemeMessage {
  func toSdkTheme() -> JessicaTheme {
    switch self {
    case .dark: return .dark
    case .light: return .light
    case .auto: return .auto
    }
  }
}

// JessicaSDKError (iOS) has no webviewOutdated/webviewRendererGone case --
// see pigeons/gfmc_api.dart's comment on GfmcErrorMessage.
private extension JessicaSDKError {
  func toMessage() -> GfmcErrorMessage {
    switch self {
    case .authFailed: return .authFailed
    case .networkError: return .networkError
    case .sessionExpired: return .sessionExpired
    case .sdkNotInitialized: return .sdkNotInitialized
    case .webViewUnavailable: return .webviewUnavailable
    case .billingUnavailable: return .billingUnavailable
    case .purchaseFailed: return .purchaseFailed
    case .verifyFailed: return .verifyFailed
    }
  }
}

private extension JessicaModule {
  func toMessage() -> GfmcModuleMessage {
    switch self {
    case .cinema: return .cinema
    case .game: return .game
    case .shop: return .shop
    }
  }
}

File: android/src/main/kotlin/com/sltr/gfmc/flutter/GfmcFlutterPlugin.kt
package com.sltr.gfmc.flutter

import android.app.Activity
import android.content.Context
import androidx.annotation.NonNull
import com.sltr.gfmc.GfmcModule
import com.sltr.gfmc.GfmcRefreshResult
import com.sltr.gfmc.GfmcSDK
import com.sltr.gfmc.GfmcSDKConfig
import com.sltr.gfmc.GfmcSDKEnv
import com.sltr.gfmc.GfmcSDKError
import com.sltr.gfmc.GfmcSDKListener
import com.sltr.gfmc.GfmcSDKTheme
import com.sltr.gfmc.GfmcSkuListener
import com.sltr.gfmc.GfmcTokenRefresher
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/**
 * Dart-facing entry point. Registered automatically via pubspec.yaml's
 * `flutter.plugin.platforms.android.pluginClass` — hosts don't touch this
 * class directly, only lib/gfmc_flutter.dart's GfmcSdk facade.
 *
 * Implements GfmcHostApi (Dart -> native calls) directly, and owns a
 * GfmcFlutterApi instance (native -> Dart calls) built once attached to an
 * engine. GfmcSDK.open()/init() need a Context; ActivityAware is what makes
 * the *current* Activity available for that, since a FlutterPlugin's own
 * `context` from onAttachedToEngine is the Application context, not
 * something GfmcSDK.open() can start an Activity from reliably on every
 * Android version.
 */
class GfmcFlutterPlugin : FlutterPlugin, ActivityAware, GfmcHostApi {

    private var applicationContext: Context? = null
    private var activity: Activity? = null
    private var flutterApi: GfmcFlutterApi? = null

    // Non-null once GfmcSDK.init() has actually been called via [initialize]
    // below.
    // Guards against onSkuSelected/refreshToken/listener callbacks firing
    // into a torn-down engine after detach.
    private var attached = false

    // GfmcSDK.getConfig() exists but is `internal` in the AAR (compile-time
    // "Cannot access ... it is internal in 'com.sltr.gfmc.GfmcSDK'" against
    // gfmc-sdk 1.2.6) -- not something a consumer module can call. We track
    // the last config passed to [initialize] ourselves instead of reading it
    // back from the SDK.
    private var lastConfig: GfmcConfigMessage? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        flutterApi = GfmcFlutterApi(binding.binaryMessenger)
        GfmcHostApi.setUp(binding.binaryMessenger, this)
        attached = true
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        attached = false
        GfmcHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // -- GfmcHostApi (Dart -> native) ----------------------------------------

    override fun initialize(config: GfmcConfigMessage) {
        val ctx = activity ?: applicationContext
            ?: throw IllegalStateException("GfmcFlutterPlugin.initialize() called before attaching to an engine")

        val sdkConfig = GfmcSDKConfig.Builder()
            .setEnvironment(config.environment.toSdkEnv())
            .setLocale(config.locale)
            .setTheme(config.theme.toSdkTheme())
            .enableLogging(config.enableLogging)
            .setConnectionTimeout(config.connectionTimeoutMs)
            .build()

        GfmcSDK.init(ctx, sdkConfig)
        GfmcSDK.setListener(ctx, sdkListener)
        GfmcSDK.setTokenRefresher(tokenRefresher)
        GfmcSDK.setSkuListener(skuListener)
        lastConfig = config
    }

    override fun open(jwt: String) {
        val ctx = activity ?: applicationContext
            ?: throw IllegalStateException("GfmcFlutterPlugin.open() called before attaching to an engine")
        GfmcSDK.open(ctx, jwt)
    }

    override fun closeMiniApp() {
        // GfmcSDK doesn't expose a "close the currently-open hub from the
        // outside" call today (closing is driven from inside the hub, via
        // its own capsule/back-press handling) -- this is intentionally a
        // documented no-op placeholder for now rather than a silent lie.
        // See jessica-sdk-android if a host-initiated close ever gets added.
    }

    override fun getVersion(): GfmcVersionMessage {
        val v = GfmcSDK.version
        return GfmcVersionMessage(
            name = v.name,
            versionCode = v.versionCode.toLong(),
            artifactVersion = v.artifactVersion,
            displayName = v.displayName,
        )
    }

    override fun getConfig(): GfmcConfigMessage? = lastConfig

    // -- GfmcSDK callbacks -> GfmcFlutterApi (native -> Dart) --------------

    private val sdkListener = object : GfmcSDKListener {
        override fun onHubReady() {
            if (attached) flutterApi?.onHubReady { }
        }
        override fun onHubClosed() {
            if (attached) flutterApi?.onHubClosed { }
        }
        override fun onError(code: GfmcSDKError, message: String) {
            if (attached) flutterApi?.onError(code.toMessage(), message) { }
        }
        override fun onPurchaseCompleted(sku: String, txId: String) {
            if (attached) flutterApi?.onPurchaseCompleted(sku, txId) { }
        }
        override fun onPurchaseFailed(reason: String) {
            if (attached) flutterApi?.onPurchaseFailed(reason) { }
        }
        override fun onModuleChanged(module: GfmcModule) {
            if (attached) flutterApi?.onModuleChanged(module.toMessage()) { }
        }
        override fun onShareRequested(url: String, title: String) {
            if (attached) flutterApi?.onShareRequested(url, title) { }
        }
    }

    private val skuListener = GfmcSkuListener { sku ->
        if (attached) flutterApi?.onSkuSelected(sku) { }
    }

    private val tokenRefresher = GfmcTokenRefresher { result: GfmcRefreshResult ->
        val api = flutterApi
        if (!attached || api == null) {
            result.fail("Flutter engine detached")
            return@GfmcTokenRefresher
        }
        api.refreshToken { refreshResult ->
            refreshResult.fold(
                onSuccess = { jwt -> result.emit(jwt) },
                onFailure = { error -> result.fail(error.message ?: "refresh failed") },
            )
        }
    }
}

// -- Enum conversions ---------------------------------------------------

private fun GfmcEnvMessage.toSdkEnv(): GfmcSDKEnv = when (this) {
    GfmcEnvMessage.PRODUCTION -> GfmcSDKEnv.PRODUCTION
    GfmcEnvMessage.SANDBOX -> GfmcSDKEnv.SANDBOX
    GfmcEnvMessage.DEV -> GfmcSDKEnv.DEV
}

private fun GfmcThemeMessage.toSdkTheme(): GfmcSDKTheme = when (this) {
    GfmcThemeMessage.DARK -> GfmcSDKTheme.DARK
    GfmcThemeMessage.LIGHT -> GfmcSDKTheme.LIGHT
    GfmcThemeMessage.AUTO -> GfmcSDKTheme.AUTO
}

private fun GfmcSDKError.toMessage(): GfmcErrorMessage = when (this) {
    GfmcSDKError.AUTH_FAILED -> GfmcErrorMessage.AUTH_FAILED
    GfmcSDKError.NETWORK_ERROR -> GfmcErrorMessage.NETWORK_ERROR
    GfmcSDKError.SESSION_EXPIRED -> GfmcErrorMessage.SESSION_EXPIRED
    GfmcSDKError.SDK_NOT_INITIALIZED -> GfmcErrorMessage.SDK_NOT_INITIALIZED
    GfmcSDKError.WEBVIEW_UNAVAILABLE -> GfmcErrorMessage.WEBVIEW_UNAVAILABLE
    GfmcSDKError.WEBVIEW_OUTDATED -> GfmcErrorMessage.WEBVIEW_OUTDATED
    GfmcSDKError.WEBVIEW_RENDERER_GONE -> GfmcErrorMessage.WEBVIEW_RENDERER_GONE
    GfmcSDKError.BILLING_UNAVAILABLE -> GfmcErrorMessage.BILLING_UNAVAILABLE
    GfmcSDKError.PURCHASE_FAILED -> GfmcErrorMessage.PURCHASE_FAILED
    GfmcSDKError.VERIFY_FAILED -> GfmcErrorMessage.VERIFY_FAILED
}

private fun GfmcModule.toMessage(): GfmcModuleMessage = when (this) {
    GfmcModule.CINEMA -> GfmcModuleMessage.CINEMA
    GfmcModule.GAME -> GfmcModuleMessage.GAME
    GfmcModule.SHOP -> GfmcModuleMessage.SHOP
}
